### PR TITLE
feat(eslint-plugin): add extension rule `dot-notation`

### PR DESCRIPTION
**Pull request by anikethsaha**
added new extended rule `dot-notation`
- added new options 

```js
          allowPrivateClassPropertyAccess: {
            tyoe: 'boolean',
            default: true,
          },
```
  - [X] tests
  - [X] docs
  - [X] typing update

Refer https://github.com/typescript-eslint/typescript-eslint/issues/789, https://github.com/typescript-eslint/typescript-eslint/issues/789#issuecomment-517374304

### CODE REVIEW

**Comment by anikethsaha** [comment]
I do think here also, the coverage might because of the optional chains

File: packages/eslint-plugin/src/rules/dot-notation.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    const allowPrivateClassPropertyAccess =
+      options.allowPrivateClassPropertyAccess;
+
+    const parserServices = getParserServices(context);
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    return {
+      MemberExpression(node: TSESTree.MemberExpression): void {
+        const objectSymbol = typeChecker.getSymbolAtLocation(
+          parserServices.esTreeNodeToTSNodeMap.get(node?.property),
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/dot-notation.ts:66]: `node` can never be nullish

```suggestion
          parserServices.esTreeNodeToTSNodeMap.get(node.property),
```

File: packages/eslint-plugin/src/rules/dot-notation.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        );
+
+        if (
+          allowPrivateClassPropertyAccess &&
+          objectSymbol?.declarations[0]?.modifiers &&
+          objectSymbol?.declarations[0]?.modifiers[0]?.kind ===
+            ts.SyntaxKind.PrivateKeyword
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/dot-notation.ts:73]: you can optional chain the index access
```suggestion
          objectSymbol?.declarations[0]?.modifiers?.[0].kind ===
            ts.SyntaxKind.PrivateKeyword
```

File: packages/eslint-plugin/src/rules/dot-notation.ts (hunk 1 above)
Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/dot-notation.ts]: done

File: packages/eslint-plugin/src/rules/dot-notation.ts (hunk 2 above)
Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/dot-notation.ts]: didnt know about that. thanks

File: packages/eslint-plugin/src/rules/dot-notation.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        },
+        additionalProperties: false,
+      },
+    ],
+    fixable: baseRule.meta.fixable,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [
+    {
+      allowPrivateClassPropertyAccess: true,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/dot-notation.ts:50]: please set it to default false - not everyone will want to allow this out of the box.

File: packages/eslint-plugin/src/rules/dot-notation.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            type: 'boolean',
+            default: true,
+          },
+          allowPattern: {
+            type: 'string',
+            default: '',
+          },
+          allowPrivateClassPropertyAccess: {
+            tyoe: 'boolean',
+            default: true,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/dot-notation.ts:39]: ditto

File: packages/eslint-plugin/docs/rules/dot-notation.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# enforce dot notation whenever possible (`dot-notation`)
+
+## Rule Details
+
+This rule extends the base [`eslint/dot-notation`](https://eslint.org/docs/rules/dot-notation) rule.
+It supports all options and features of the base rule.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/dot-notation.md:6]:
```suggestion
It adds support for optionally ignoring computed `private` member access.
```

File: packages/eslint-plugin/docs/rules/dot-notation.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  "dot-notation": "off",
+  "@typescript-eslint/dot-notation": ["error"]
+}
+```
+
+## Options
+
+New options
+
+- `allowPrivateClassPropertyAccess`
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/dot-notation.md:22]: Based on some recent standardisation of extension rule docs

````suggestion
See [`eslint/dot-notation`](https://eslint.org/docs/rules/dot-notation#options) options.
This rule adds the following options:

```ts
interface Options extends BaseDotNotationOptions {
  allowPrivateClassPropertyAccess?: boolean;
}

const defaultOptions: Options = {
  ...baseDotNotationDefaultOptions,
  allowPrivateClassPropertyAccess: false,
};
```

### `allowPrivateClassPropertyAccess`
````

File: packages/eslint-plugin/docs/rules/dot-notation.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+}
+```
+
+```cjson
+{
+  "allowPrivateClassPropertyAccess": true,
+}
+```
+
+See [`eslint/dot-notation` options](https://eslint.org/docs/rules/dot-notation#options).
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/dot-notation.md:38]: Replace these lines with an example of valid code with the option set to `true`

```suggestion
```

**Comment by anikethsaha** [comment]
thanks a lot for quick reviews and all those helping changes.
those are always helpful 👍

**Pull request merged**